File: Python/2-Medium/00002._两数相加/00002._两数相加.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def addTwoNumbers(l1: ListNode, l2: ListNode) -> ListNode:
    # 哨兵节点，简化边界情况处理
    dummy_head = ListNode(0)
    current = dummy_head
    carry = 0
    
    # 同时遍历两个链表
    while l1 or l2:
        # 如果链表已经遍历完，就使用0作为值
        x = l1.val if l1 else 0
        y = l2.val if l2 else 0
        
        sum = carry + x + y
        carry = sum // 10  # 更新进位
        current.next = ListNode(sum % 10)  # 创建新节点，加入到结果链表
        current = current.next  # 移动指针
        
        # 移动l1和l2指针
        if l1:
            l1 = l1.next
        if l2:
            l2 = l2.next
    
    # 检查最后的进位
    if carry > 0:
        current.next = ListNode(carry)
    
    return dummy_head.next  # 返回哨兵节点的下一个节点，即结果链表的头节点

# Helper function to convert list to ListNode
def list_to_listnode(numbers):
    dummy_root = ListNode(0)
    ptr = dummy_root
    for number in numbers:
        ptr.next = ListNode(number)
        ptr = ptr.next

    ptr = dummy_root.next
    return ptr

# Helper function to print ListNode
def print_listnode(ln: ListNode):
    while ln:
        print(ln.val, end=" -> " if ln.next else "")
        ln = ln.next

# Test cases
l1 = list_to_listnode([2, 4, 3])
l2 = list_to_listnode([5, 6, 4])

result = addTwoNumbers(l1, l2)

# Output the result
print_listnode(result)

File: C++/2-Medium/00002._两数相加/00002._两数相加.cpp
#include <iostream>

// 定义链表节点结构体
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 实现两数相加函数
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummyHead = new ListNode(0); // 创建哑节点，简化头节点的插入操作
        ListNode *p = l1, *q = l2, *curr = dummyHead;
        int carry = 0; // 初始化进位为0
        while (p != nullptr || q != nullptr) {
            int x = (p != nullptr) ? p->val : 0;
            int y = (q != nullptr) ? q->val : 0;
            int sum = carry + x + y;
            carry = sum / 10; // 更新进位
            curr->next = new ListNode(sum % 10); // 创建新节点，加入到结果链表
            curr = curr->next;
            if (p != nullptr) p = p->next;
            if (q != nullptr) q = q->next;
        }
        if (carry > 0) {
            curr->next = new ListNode(carry); // 处理最后的进位
        }
        ListNode *result = dummyHead->next; // 获取结果链表的头节点
        delete dummyHead; // 释放哑节点
        return result;
    }
};

// 辅助函数，用于创建和打印链表，仅为了完整演示，实际提交时不需要
ListNode* createList(std::initializer_list<int> vals) {
    ListNode *dummyHead = new ListNode(0);
    ListNode *curr = dummyHead;
    for (int val : vals) {
        curr->next = new ListNode(val);
        curr = curr->next;
    }
    ListNode *result = dummyHead->next;
    delete dummyHead;
    return result;
}

void printList(ListNode *node) {
    while (node != nullptr) {
        std::cout << node->val << " -> ";
        node = node->next;
    }
    std::cout << "nullptr" << std::endl;
}

// 主函数，仅为了演示如何使用，实际提交时不需要
int main() {
    Solution solution;
    ListNode *l1 = createList({2, 4, 3});
    ListNode *l2 = createList({5, 6, 4});
    ListNode *result = solution.addTwoNumbers(l1, l2);
    printList(result);
    // 应该输出：7 -> 0 -> 8 -> nullptr
    return 0;
}